### PR TITLE
Feature and Admin Upgrade/Overhaul

### LINKED ISSUE #3: More Condtions Needed for Different Dungeon/Raid Type Scaling
Assuming the config file is set at the following default values of:

Solocraft.Dungeon = 5 
Solocraft.Heroic = 10
Solocraft.Raid25 = 25
Solocraft.Raid40 = 40

Vanilla Dungeon/Raid Behavior:

5-man dungeon scaling = offset 5 - Correct!
10-man raid scaling (Upper Blackrock Spire) = offset 5 - Incorrect! Should offset to 10
20-man raid scaling (AQ20 and Zul'Gurub) = offset 40 - Incorrect! Should offset to 20
40-man raid scaling = offset 40 - Correct!

TBC Dungeon/Raid Behavior:

5-man dungeon scaling = offset 5 - Correct!
5-man Heroic dungeon scaling = offset 10 - Correct because the config is set to 10 for Heroic, but realistically it should still be 5 by default, but we don't have code/config that differentiates between 5 man heroic dungeon and 10 man heroic raid.
10-man Raid scaling = offset 40 - Incorrect! Should offset to 10
25-man Raid scaling = offset 40 - Incorrect! Should offset to 25

Wotlk Dungeon/Raid Behavior:
5-man dungeon scaling = offset 5 - Correct!
5-man Heroic dungeon scaling = offset 10 - Correct because the config is set to 10 for Heroic, but realistically it should still be 5 by default, but we don't have code/config that differentiates between 5 man heroic dungeon and 10 man heroic raid.
10-man Raid scaling = offset 40 - Incorrect! Should offset to 10
25-man Raid scaling  offset 25 - Correct!
10-man Heroic Raid scaling = offset 10 - Correct because the config is set to 10 for Heroic, but we don't have code/config that differentiates between 5 man heroic dungeon and 10 man heroic raid.
25-man Heroic Raid scaling = offset 25 - Correct!

Based on the above data, the following issues can be deduced:

1. In General:
a) Solocraft.Heroic = 10 seems to apply to BOTH 5 man Heroic dungeons and 10 man Heroic raids.
b) Solocraft.Raid40 = 40 seems to apply to all TBC raids and Wotlk 10 man (non-Heroic ) raids. Also applies to Vanilla 20 man raids.

1. Vanilla:
a) No conditons to differentiate between 20 man and 40 man raids.
b) UBRS being a special 10-man mini raid is still being treated as a 5 man dungeon.

2. TBC:
a) 10 man and 25 man raids are still being treated as 40 man raids. Missing conditions/flags for TBC raids since they have changed in max player number since Vanilla.

3. Wotlk:
a) 10 man raids are still being treated as 40 man raids. Again missing conditions/flags.
b) Solocraft.Raid25 = 25 only applies to 25 man raids in Wotlk but not TBC, and it applies to BOTH 25 man Heroic and non-Heroic raids.

Improvements to the Solocraft code and config are required to meet the following complete set of conditions for Vanilla/TBC/Wotlk Dungeon & Raid content:

Solocraft.Dungeon = 5 (No changes. Keep as it is for all vanilla/tbc/wotlk 5 man dungeons)
Solocraft.Heroic.Dungeon = 5 (Add this to take care of ONLY 5 man Heroic dungeons in TBC and Wotlk)
Solocraft.Raid10 = 10 (Add this to take care of ONLY the 10 man non-Heroic raids in TBC and Wotlk, as well as UBRS in Vanilla)
Solocraft.Heroic.Raid10 = 10 (Add this to take care of ONLY the 10 man Heroic raids in Wotlk)
Solocraft.Raid20 = 20 (Add this to take care of ONLY the 20 man raids in Vanilla.)
Solocraft.Raid25 = 25 (Modify this so it takes care of ONLY the 25 man non-Heroic raids in TBC and Wotlk.)
Solocraft.Heroic.Raid25 = 25 (Add this to take care of ONLY the 25 man Heroic raids in Wotlk)
Solocraft.Raid40 = 40 (Modify this so it takes care of ONLY the 40 man raids in Vanilla)

### LINKED ISSUE #9: Logging out while in a dungeon and logging back in does not rebuff the character
While using Solocraft I noticed that if for some reason I logout while being in a dungeon/instance and then relog back in I do not get the buff from this module.  To add insult to injury if I leave the instance my character gets debuffed again.  

Looking at the code and doing some minor testing it appears it debuffs your character on login and then turns around and tries to add the buff to your character.  I assume this is because of mapchange being called a few times on login.

I noticed it runs through this code both times:

```
            // Adjust player stats
            _unitDifficulty[player->GetGUID()] = difficulty;
            for (int32 i = STAT_STRENGTH; i < MAX_STATS; ++i)
            {
                // Buff the player
                player->HandleStatModifier(UnitMods(UNIT_MOD_STAT_START + i), TOTAL_PCT, float(difficulty * 100), true);
            }

            // Set player health
            player->SetFullHealth();
            if (player->getPowerType() == POWER_MANA)
            {
                // Buff the player's health
                player->SetPower(POWER_MANA, player->GetMaxPower(POWER_MANA));
            }
```

The first time it passes through the difficulty is 1.  The second time it uses the set dungeon difficulty but it fails to apply the buff.  I think it works only on a mapchange because the character goes through some sort of refresh.   The HandleStatModifier and the SetFullHealth methods don't seem to do the refresh alone.  I have been trying to fix this but can't find any documentation or any examples in the Azerothcore source code that has worked so far.  

I have tried these methods with little (saw some stats refresh but not mana or hp changes) or no changes:
ApplyStatPercentBuffMod
UpdateAllStats
UpdateMaxHealth

If anyone who reads this has any clues or comments I would appreciate it.

### PR DESCRIPTION
**Pull request by LannyE**
This is a large overhaul of the Solocraft module.

Features and Admin Controls added:

-Groups now work. Difficulty is now distributed across group size
-Debuff added for groups trying to exceed the difficulty modifier
-Debuff can be enabled and disabled
-Spellpower is now scaled based on character level and modifier
-Stat modifier can now be changed in config to reel in melee damage
-Individual modification of each dungeon/raid difficulty
-Max level cap to keep high level characters from getting buffed
-Max level cap modifiable in config for each dungeon/raid
-Dungeon/Raid types still used as a catch all if not found as an
individual instance in the config

A database table is included to manage groups and the difficulty
modifier that each group member has applied to them. It also
deals with server restarts and players having to log out and log
back in to try to correct stats. The table is required.

Updated README.md file with current instructions and config file name fix.

Fixes #9
closes #3

Testing:

Debuff:
-Character in dungeon with full difficulty modifier applied while not in a group.
-Invite another character to the group.
-Have new group member enter same dungeon.
-Debuff will be applied because of the group member already in the dungeon
has the full difficulty modifier applied.
-Have both/all characters leave dungeon and re-enter it to have the modifier
balanced across the current members of the group.

-Max Level:
-Solocraft.Max.Level.Diff and the Dungeon individual level are added together
to declare the max level a character can be to get the difficulty modifier
applied for the dungeon.
-Use a character whose level is higher than the max level and enter dungeon.
-Character should not receive the difficulty modifier buff for the dungeon.

-Spellpower and Stats modifier adjustments:
-Change modifier in the config.
-Enter dungeon with character and notice stats and/or spellpower numbers
changed from previous entrance into the dungeon.

### CODE REVIEW

**Comment by Helias** [comment]
Now this has conflicts with the other PR that I just merged :sweat_smile: 
Solve conflicts please

**Comment by LannyE** [comment]
This is an adventure.   You can definitely tell I am new to this.  Hopefully they have been resolved correctly.

**Comment by Helias** [comment]
great job!

File: src/Solocraft.cpp [hunk 1]
```diff
@@ -77,106 +490,722 @@ class solocraft_player_instance_handler : public PlayerScript {
...
+				}	
+				else if (map->GetId() == 668) {
+					difficulty = D668H;
+				}				
+				//WOTLK 10 Man Heroic
+				else if (map->GetId() == 649) {
+					difficulty = D649H10;
+				}					
+			else //Catch All	-----------10 Heroics
                 difficulty = D10;
```
Comment (Helias) [inline src/Solocraft.cpp:646]: conver this into a switch-case or better, into an hashmap pls

File: src/Solocraft.cpp [hunk 2]
```diff
@@ -77,106 +490,722 @@ class solocraft_player_instance_handler : public PlayerScript {
...
             else if (map->IsDungeon()) {
                 difficulty = D5;
             }
+            else if (map->IsRaid()) {
+                difficulty = D40;
+            }			
         }
-        return difficulty;
+        return difficulty; //return the specific dungeon's difficulty offset
     }
```
Comment (Helias) [inline src/Solocraft.cpp:885]: even here, a hashmap is a lot better, you can even improve the configs variables

File: src/Solocraft.cpp [hunk 3]
```diff
@@ -77,106 +490,722 @@ class solocraft_player_instance_handler : public PlayerScript {
...
+				dunLevel = M658; break;
+			case 668: 
+				dunLevel = M668; break;
+			case 724: 
+				dunLevel = M724; break;		
+			default:
+				dunLevel = SolocraftDungeonLevel;
+		}
+        return dunLevel; //return the specific dungeon's level 
+	}
```
Comment (Helias) [inline src/Solocraft.cpp:1044]: if you create a hashmap you don't need all of this code

**Review by Helias — CHANGES_REQUESTED** [review]

File: src/Solocraft.cpp (hunk 3 above)
Comment (Helias) [reply src/Solocraft.cpp:1044]: exapmle:
```C++
#include <unordered_map>
std::unordered_map<uint32, uint32> dungeons = {
  { 33, sConfigMgr->GetIntDefault("Solocraft.ShadowfangKeep.Level", 15) }
};
```

and to use it as dunLevel you can just use
```C++
dunLevel = deungeons[map->GetId()];
```
or better, `return  deungeons[map->GetId()];`

**Review by Helias — CHANGES_REQUESTED** [review]

**Comment by LannyE** [comment]
Thanks for showing me the hashmap option.  That greatly reduced the number of lines of code.   I made the changes and did some testing.  It seems to be working fine.

**Review by Helias — APPROVED** [review]

**Comment by Helias** [comment]
It is much better, doesn't it? :grin: 
Good job!

**Comment by Helias** [comment]
I improved the comments next-to the dungeons.
Anyways, a hints for the future, convert `tabs` into `spaces` :wink: 

I will wait the build and merge it if the bild succeed.

**Comment by LannyE** [comment]
The changes looks good.  Thank you for all your help. I appreciate your patience while I navigate this learning curve.  I haven't programmed in C++ for about 20+ years and never used Git (which is obvious).  On a side note: I also have a bunch of database changes for the lunar event and a new module I made based off a script I found.  Once I get some free time I will take a shot at uploading those.

**Comment by LannyE** [comment]
Also this push should fix issue #3.  I used the testing done in that issue to form the logic and config settings.

**Comment by Helias** [comment]
About #3 I edited the description of the PR so when we merge the PR the issue will be closed, for the module and lunar make a PR and create a repo with your module ;-)

**Pull request merged**